### PR TITLE
Allow CSV ingest to create new shots.

### LINKED ISSUE #24: AY-6731_CSV publishing nonexistent hierarchy
### Please describe the enhancement you have in mind and explain what the current shortcomings are?

The CSV workflow currently can't create new hierarchies in the AY project. We'll need to use Editorial context-based hierarchy data publishing instead. This may not require a lot of work, as  recently developed a helper function for this during Circuit development.

The process of CSV ingest after the enhancement will need to do following:
Check if the mentioned folder path on row exists
If it does not exists then it will need to create Hierarchical data hierarchyContext on context for it.
It will also need to add bool key into instance data with newHierarchyIntegration set to True
Then we will need to figure out what would be the best family to add into ExtractHierarchyToAYON and IntegrateHierarchyToFtrack  and even perhaps itno Shotgrid and other PM addons. But there might be better way of doing this since  was mentioning that new Publishing workflow allows for plugin to demand other plugins downstream.

### How would you imagine the implementation of the enhancemenent?
_No response_

### Describe alternatives you've considered:
_No response_

### Additional context:
()
_(might be a private channel)_

_This issue was automatically created from Clickup ticket [AY-6731](https://app.clickup.com/t/86c0a6d6w)_

### PR DESCRIPTION
**Pull request by robin-ynput**
## Changelog Description

resolve #24

#### New features:
* Allow CSV ingest to create new shots for non-existent folder paths
* Provide additional server settings for Folder and Task creation

image

## Additional review information

**Current implementation has some limitations**:
* Created shot are pretty much blank as we do not have any OTIO/range/resolution info from the CSV file

## Testing notes:
This needs to be tested along those changes in `ayon_core`: https://github.com/ynput/ayon-core/pull/1010

1. Create a new package and upload it to your AYON server
2. In CSV settings, enable missing folder creation (`ayon+settings://traypublisher/create/IngestCSV/folder_creation_config`)
3. Prepare a CSV file with Folder Path that do not exist yet
4. Ingest the CSV file through the `publisher` option in traypublisher
5. Ensure data is publish as expected and new blank shot(s) have been created

### CODE REVIEW

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py [hunk 1]
```diff
@@ -482,6 +526,22 @@ def _get_data_from_csv(
...
+                product_item.has_promised_context = True
+                product_item.task_type = None
+                try:
+                    product_item.parents = self._validate_parents(
+                        project_name,
+                        product_item
+                    )
+                except ValueError:
+                    raise CreatorError(
+                        f"Parent context must exists for new shots: {folder_path}"
```
Comment (robin-ynput) [inline client/ayon_traypublisher/plugins/create/create_csv_ingest.py:540]: The code expects the parent folder to newly-created shots to exists.
I thought it was to ambigious to assume to folder type(s) for the whole hierarchy otherwise. 

Is that a good behavior ?

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py [hunk 2]
```diff
@@ -835,6 +895,23 @@ def _create_instances_from_csv_data(self, csv_dir: str, filename: str):
...
+                        "hierarchy": hierarchy,
+                        "parents": product_item.parents
+                    }
+                )
+                # TODO create new task from provided task name
+#                if product_item.task_name:
+#                    instance_data["tasks"] = {
+#                        "name": product_item.task_name,
+#                        "type": "Generic"
+#                    }
```
Comment (robin-ynput) [inline client/ayon_traypublisher/plugins/create/create_csv_ingest.py:913]: I wasn't sure what the `tasks` data should be here in order for `Extract Hierarchy to AYON` to create a new task along the new shots. From other code I thought the `dict` was correct but it looks like it's not.

traypublisher_task_dict

Unless it's an easy one and I missed something, I think this needs to be fixed in its own PR.

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py [hunk 3]
```diff
@@ -835,6 +895,23 @@ def _create_instances_from_csv_data(self, csv_dir: str, filename: str):
                 "prepared_data_for_repres": []
             }
 
+            if product_item.has_promised_context:
+                hierarchy, _ = folder_path.rsplit("/",1)
+                families.append("csv_ingest_shot")
```
Comment (robin-ynput) [inline client/ayon_traypublisher/plugins/create/create_csv_ingest.py:900]: happy to discuss this one, we kind of need something to trigger the `collect hierarchy` and `extract hierarchy to AYON` plugins but cannot use `clip` nor `shot` cause we do not have any OTIO data here.

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py [hunk 4]
```diff
@@ -835,6 +895,23 @@ def _create_instances_from_csv_data(self, csv_dir: str, filename: str):
                 "prepared_data_for_repres": []
             }
 
+            if product_item.has_promised_context:
+                hierarchy, _ = folder_path.rsplit("/",1)
```
Comment (moonyuet) [inline client/ayon_traypublisher/plugins/create/create_csv_ingest.py:899]:
```suggestion
                hierarchy, _ = folder_path.rsplit("/", 1)
```
some cosmetic

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py (hunk 2 above)
Comment (moonyuet) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:913]: I believe the tasks data should be part of the instances. But guess it isn't there as somehow the commented code is a bit hard-coded and it is not updating the dictionary data with name and type per task.

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py [hunk 5]
```diff
@@ -388,6 +390,54 @@ def _resolve_repre_path(
...
+        """
+        parent_folder_path, folder_name = product_item.folder_path.rsplit("/", 1)
+        if not parent_folder_path:
+            return []
+
+        folder_data = ayon_api.get_folder_by_path(
+            project_name,
+            parent_folder_path,
+            fields=("folderType",)
+        )
```
Comment (moonyuet) [inline client/ayon_traypublisher/plugins/create/create_csv_ingest.py:415]: Is it possible to do this that way?(I suppose `product_item.folder_path` is storing `folder_path`)
```suggestion
        parent_folder_path = product_item.folder_path
        folder_data = ayon_api.get_folder_by_path(
            project_name,
            parent_folder_path,
            fields=("folderType")
        )
```

Comment (BigRoy) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:415]: The recommendation above wouldn't work, because it doesn't make `fields` a `tuple` anymore.

```suggestion
        parent_folder_path = product_item.folder_path
        folder_data = ayon_api.get_folder_by_path(
            project_name,
            parent_folder_path,
            fields={"folderType"}
        )
```

However, I believe `ayon_api` internally makes `fields` a `set` anyways, so we can just pass it a set. I've updated Kayla's example here.

I'm not sure why @moonyuet is recommending that however, because we're then also querying a different folder, because we're now not querying the parent folder anymore? Which I suspect was intentionally done.

So the only change would be changing the tuple to a set, which is just cosmetics at this point:

```suggestion
        parent_folder_path, folder_name = product_item.folder_path.rsplit("/", 1)
        if not parent_folder_path:
            return []

        folder_data = ayon_api.get_folder_by_path(
            project_name,
            parent_folder_path,
            fields={"folderType"}
        )
```

Comment (moonyuet) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:415]: > I'm not sure why @moonyuet is recommending that however, because we're then also querying a different folder, because we're now not querying the parent folder anymore? Which I suspect was intentionally done.

I misunderstood the comment here https://github.com/ynput/ayon-unreal/blob/eedb2beec8d0af53b27f93831431956a9e8d7e9d/client/ayon_unreal/api/pipeline.py#L610-L622, so ignore me then.

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py [hunk 6]
```diff
@@ -388,6 +390,54 @@ def _resolve_repre_path(
...
+                project_name,
+                f"/{'/'.join(inspected)}",
+                fields=("folderType",)
+            )
+            parent_data.append({
+                "folder_type": folder_data["folderType"],
+                "entity_name": key
+            })
+
+        return parent_data
```
Comment (iLLiCiTiT) [inline client/ayon_traypublisher/plugins/create/create_csv_ingest.py:438]: Simpified, with single call to server.
```suggestion
        parent_folder_names = product_item.folder_path.lstrip("/").split("/")
        # Rename name of folder itself
        parent_folder_names.pop(-1)
        if not parent_folder_names:
            return []
    
        parent_paths = []
        parent_path = ""
        for name in parent_folder_names:
            path = f"{parent_path}/{name}"
            parent_paths.append(path)
            parent_path = path
    
        folders_by_path = {
            folder["path"]: folder
            for folder in ayon_api.get_folders(
                project_name,
                folder_paths=parent_paths,
                fields={"folderType", "path"}
            )
        }
        parent_data = []
        for path in parent_paths:
            folder_entity = folders_by_path.get(path)
            if not folder_entity:
                raise ValueError(f"Unknown parent folder {path}")
            name = path.rsplit("/", 1)[-1]
            parent_data.append({
                "folder_type": folder_entity["folderType"],
                "entity_name": name
            })
        return parent_data
```

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py (hunk 1 above)
Comment (iLLiCiTiT) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:540]: I don't think it is. We don't have any logic to define folder types of parents, it is right now hardcoded (as far as I know), but they usually don't exist on first editorial publish.

Comment (jakubjezek001) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:540]: This is quite challenging. Non-existent parent types should be defined in the settings. If they match, they can be resolved. If not, let's set them as the default folder type.

  

The mapping for folder names to types could use a regex pattern input. The settings for this mapping should be added to the traypublisher CSV editorial here: `ayon+settings://traypublisher/create/IngestCSV`.

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py (hunk 2 above)
Comment (jakubjezek001) [inline client/ayon_traypublisher/plugins/create/create_csv_ingest.py:913]:
```suggestion
                if product_item.task_name:
                    instance_data["tasks"] = [
	                    {
	                        "name": product_item.task_name,
	                        "type": "Generic"
	                    }
                    ]
```

Comment (jakubjezek001) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:913]:
```suggestion
                if product_item.task_name:
                    tasks = instance_data.setdefault("tasks", [])
                    tasks.append(
                        {
                            "name": product_item.task_name,
                            "type": "Generic"
                        }
                    )
```

Or since I am not sure if tasks could be added in batch we could perhaps do it this way.

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py (hunk 5 above)
Comment (robin-ynput) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py]: I went with @iLLiCiTiT recommendation which leverage `ayon_api.get_folders`.

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py (hunk 2 above)
Comment (robin-ynput) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py]: I'm confused now because if I put in a list, it fails here: https://github.com/ynput/ayon-core/blob/6a07de6eb904c139f6d346fd6f2a7d5042274c71/client/ayon_core/plugins/publish/collect_anatomy_instance_data.py#L497
```
AttributeError: 'list' object has no attribute 'get'
```

Comment (iLLiCiTiT) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py]: I think it is task info by name
```
"tasks": {
    "Modeling": {
        "type": "modeling",
    },
    "Animation": {
        "type": "animation",
    },
    ...
}
```

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py [hunk 7]
```diff
@@ -388,6 +391,66 @@ def _resolve_repre_path(
...
+                    if re.match(folder_setting["regex"], name):
+                        folder_type = folder_setting["folder_type"]
+                        break
+                else:
+                    folder_type = "Folder"  # default
+
+            parent_data.append({
+                "folder_type": folder_type,
+                "entity_name": name
+            })
```
Comment (iLLiCiTiT) [inline client/ayon_traypublisher/plugins/create/create_csv_ingest.py:449]: I would keep it without hardcoded folder type if possible (so the integrator in ayon-core can handle it). OR add to settings `default_folder_type` attribute which would be used in that case.
```suggestion
            folder_type = None
            if folder_entity:
                folder_type = folder_entity["folderType"]

            # Define folder type from settings.
            else:
                for folder_setting in self.folder_creation_config["folder_type_regexes"]:
                    if re.match(folder_setting["regex"], name):
                        folder_type = folder_setting["folder_type"]
                        break

            item = {
                "entity_name": name
            }
            if folder_type:
                item["folder_type"] = folder_type,
            parent_data.append(item)
```

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py (hunk 3 above)
Comment (iLLiCiTiT) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:920]: Why we can't use `shot`?

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py (hunk 1 above)
Comment (jakubjezek001) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:560]: @robin-ynput it seems you are having very old server - update it please. The list items are now looking different for some time. This is how they look now:  image

 Anyway, the folder types will better be defined as enumerator of available folder types from project anatomy. I will commit the recommendations here f199802fe994c90dc54f74f59cc4fc30852b2451

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py (hunk 3 above)
Comment (jakubjezek001) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:920]: Shot is expecting otioClip at `ayon-traypublisher\client\ayon_traypublisher\plugins\publish\collect_shot_instances.py`

Comment (jakubjezek001) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:920]: https://github.com/ynput/ayon-core/pull/1010#issuecomment-2479069045

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py [hunk 8]
```diff
@@ -835,6 +915,26 @@ def _create_instances_from_csv_data(self, csv_dir: str, filename: str):
...
+                )
+                # TODO create new task from provided task name
+                #if product_item.task_name:
+                #    tasks = instance_data.setdefault("tasks", [])
+                #    tasks.append(
+                #        {
+                #            "name": product_item.task_name,
+                #            "type": "Generic"
+                #        }
+                #    )
```
[the remark this answers is not in the document]
Comment (iLLiCiTiT) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:936]: I would skip the `update`, but yes.
```
                if product_item.task_name:
                    tasks = instance_data.setdefault("tasks", {})
                    tasks[product_item.task_name] = {
                        "type": "Generic"
                    }
```

[the remark this answers is not in the document]
Comment (robin-ynput) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py:936]: Yeah, I guess I'll report some settings options so the user can control the Task type from his own needs.

[the remark this answers is not in the document]
Comment (robin-ynput) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py]: Adjusted with some task settings options.

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py (hunk 2 above)
Comment (robin-ynput) [reply client/ayon_traypublisher/plugins/create/create_csv_ingest.py]: That was it, thanks a lot for that !

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py [hunk 9]
```diff
@@ -388,6 +391,68 @@ def _resolve_repre_path(
 
         return filepath
 
+    def _validate_parents(self, project_name: str, product_item: ProductItem) -> list:
```
Comment (jakubjezek001) [inline client/ayon_traypublisher/plugins/create/create_csv_ingest.py:394]: `_validate_parents` is a misleading name. One could expect that the parents are only checking for existence. This function is also creating parent data. How about to rename it to something like `_ensure_parents_exists` or something like it.

File: client/ayon_traypublisher/plugins/create/create_csv_ingest.py [hunk 10]
```diff
@@ -388,6 +391,68 @@ def _resolve_repre_path(
...
+            # Folder exists, retrieve data from existing.
+            if folder_entity:
+                folder_type = folder_entity["folderType"]
+
+            # Define folder type from settings.
+            else:
+                for folder_setting in self.folder_creation_config["folder_type_regexes"]:
+                    if re.match(folder_setting["regex"], name):
+                        folder_type = folder_setting["folder_type"]
+                        break
```
Comment (jakubjezek001) [inline client/ayon_traypublisher/plugins/create/create_csv_ingest.py:443]: For some reason following Path column data is producing following hierarchy with Folder type decorated in `<>`

- `/shots/mw_110_01_0060`

```
/shots <Shot>
└── mw_110_01_0060 <Shot>
```

I believe that since I am using default settings with 
image The types should be defined as:

```
/shots <Shot>
└── mw_110_01_0060 <Folder>
```

**Review by jakubjezek001 — APPROVED** [review]
My tests were successful. I was using CLI and interactive. All was working as expected. The folder type issue is not a blocker but if @robin-ynput will fix it, it would be benefitial.

**Comment by robin-ynput** [comment]
> My tests were successful. I was using CLI and interactive. All was working as expected. The folder type issue is not a blocker but if @robin-ynput will fix it, it would be benefitial.

Thanks for the review, I managed to fix the folder type issue but I had to edit the harcoded value in `ayon-core` to make it a default instead:
https://github.com/ynput/ayon-core/pull/1010/files#diff-e23278ce4a2b0b416005af0e9b9211e7c2e206320d253284272f156987f42e52R46

Can you please confirm this is OK for you ?

**Comment by jakubjezek001** [comment]
> Thanks for the review, I managed to fix the folder type issue but I had to edit the harcoded value in `ayon-core` to make it a default instead: [ynput/ayon-core#1010 (files)](https://github.com/ynput/ayon-core/pull/1010/files#diff-e23278ce4a2b0b416005af0e9b9211e7c2e206320d253284272f156987f42e52R46)
> 
> Can you please confirm this is OK for you ?

Thanks for the suggestion. I agree that it makes sense to keep it there for the context of `shot_data`. It makes sense to have the default folder type as **Shot.** Maybe we should use the same approach as the default task type and have a default folder type too.

image

**Comment by robin-ynput** [comment]
> > Thanks for the review, I managed to fix the folder type issue but I had to edit the harcoded value in `ayon-core` to make it a default instead: [ynput/ayon-core#1010 (files)](https://github.com/ynput/ayon-core/pull/1010/files#diff-e23278ce4a2b0b416005af0e9b9211e7c2e206320d253284272f156987f42e52R46)
> > Can you please confirm this is OK for you ?
> 
> Thanks for the suggestion. I agree that it makes sense to keep it there for the context of `shot_data`. It makes sense to have the default folder type as **Shot.** Maybe we should use the same approach as the default task type and have a default folder type too.
> 
> image

Yes that's exactly what in there for now:
image

The `Shot` harcoded change I mentioned was for `ayon-core` (which is now a default instead of hardcoded).
